server/src/main/java/org/apache/druid/client/CachingClusteredClient.java

```diff
@@ -289,14 +289,17 @@ private ClusterQueryResult(Sequence<T> sequence, int numQueryServers)
       this.query = queryPlus.getQuery();
       this.toolChest = warehouse.getToolChest(query);
       this.strategy = toolChest.getCacheStrategy(query);
+      this.dataSourceAnalysis = DataSourceAnalysis.forDataSource(query.getDataSource());
 
-      this.useCache = CacheUtil.isUseSegmentCache(query, strategy, cacheConfig, CacheUtil.ServerType.BROKER);
-      this.populateCache = CacheUtil.isPopulateSegmentCache(query, strategy, cacheConfig, CacheUtil.ServerType.BROKER);
+      // Broker join caching is disabled - https://github.com/apache/druid/issues/10444
+      this.useCache = CacheUtil.isUseSegmentCache(query, strategy, cacheConfig, CacheUtil.ServerType.BROKER)
+                      && !dataSourceAnalysis.isJoin();
```

Comment:
I think this check should be not here, but inside `CacheUtil.isUseSegmentCache()`. The reason we pass the `ServerType` in `isUseSegmentCache()` is that we want to make different decisions based on the `ServerType`. We are disabling the cache for join on brokers, so it should be in there.